modules/broker-auth/src/main/java/io/ballerina/messaging/broker/auth/authorization/authorizer/rdbms/scope/dao/impl/AuthScopeRdbmsDao.java

```diff
@@ -0,0 +1,160 @@
+import io.ballerina.messaging.broker.auth.exception.BrokerAuthServerException;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import javax.sql.DataSource;
+
+/**
+ * Class implements {@link AuthScopeDao} to provide database functionality to manage scopes.
+ */
+public class AuthScopeRdbmsDao extends AuthScopeDao {
+
+    public AuthScopeRdbmsDao(DataSource dataSource) {
+        super(dataSource);
+    }
+
+    @Override
+    public AuthScope read(String scopeName) throws BrokerAuthServerException {
+        Set<String> userGroups = new HashSet<>();
+        String scopeId = null;
+        Connection connection = null;
+        PreparedStatement statement = null;
+        ResultSet resultSet = null;
+        try {
+            connection = getConnection();
+            statement = connection.prepareStatement(RdbmsConstants.PS_SELECT_AUTH_SCOPE);
+            statement.setString(1, scopeName);
+            resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                if (Objects.isNull(scopeId)) {
+                    scopeId = resultSet.getString(1);
+                }
+                String userGroup = resultSet.getString(2);
+                if (Objects.nonNull(userGroup)) {
+                    userGroups.add(userGroup);
+                }
+            }
+        } catch (SQLException e) {
+            throw new BrokerAuthServerException("Error occurred while retrieving scope for name : " + scopeName, e);
+        } finally {
+            close(connection, statement, resultSet);
+        }
+
+        if (Objects.nonNull(scopeId)) {
+            return new AuthScope(scopeName, userGroups);
+        }
+        return null;
+    }
+
+    @Override
+    public List<AuthScope> readAll() throws BrokerAuthServerException {
+        Map<String, AuthScope> authScopes = new HashMap<>();
+        Connection connection = null;
+        PreparedStatement statement = null;
+        ResultSet resultSet = null;
+        try {
+            connection = getConnection();
+            statement = connection.prepareStatement(RdbmsConstants.PS_SELECT_ALL_AUTH_SCOPES);
+            resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                String scopeName = resultSet.getString(1);
+                AuthScope authScope = authScopes.get(scopeName);
+                if (Objects.isNull(authScope)) {
+                    authScope = new AuthScope(scopeName, new HashSet<>());
+                    authScopes.put(scopeName, authScope);
+                }
+                if (Objects.nonNull(resultSet.getString(2))) {
+                    authScope.getAuthorizedUserGroups().add(resultSet.getString(2));
```

Comment:
Shall we add a method AuthScope#addUserGroup instead of retrieving the underlying set and appending it?

Reply:
Fixed in d43ea86